src/proxy/pipeline/upstream.rs

```diff
@@ -1,232 +1,295 @@
-use tokio::time::{Duration, Instant};
-
-use crate::cluster::Endpoint;
-use crate::filters::{manager::SharedFilterManager, Filter, WriteContext};
-use crate::proxy::sessions::error::Error;
-use crate::proxy::sessions::metrics::Metrics;
-use crate::utils::debug;
-
-type Result<T> = std::result::Result<T, Error>;
-
-/// Session encapsulates a UDP stream session
-pub struct Session {
-    log: Logger,
-    metrics: Metrics,
-    filter_manager: SharedFilterManager,
-    /// created_at is time at which the session was created
-    created_at: Instant,
-    socket: Arc<UdpSocket>,
-    /// dest is where to send data to
-    dest: Endpoint,
-    /// from is the original sender
-    from: SocketAddr,
-    /// The time at which the session is considered expired and can be removed.
-    expiration: Arc<AtomicU64>,
-    /// a channel to broadcast on if we are shutting down this Session
-    shutdown_tx: watch::Sender<()>,
+use tokio::{
+    net::UdpSocket,
+    sync::mpsc,
+    task::JoinHandle,
+    time::{Duration, Instant},
+};
+
+use crate::{
+    cluster::Endpoint,
+    proxy::ShutdownRx,
+    filters::{manager::SharedFilterManager, Filter, WriteContext},
+    utils::debug,
+};
+
+pub use self::{error::Error, metrics::Metrics, packet::UpstreamPacket};
+
+type UpstreamMap = dashmap::DashMap<(SocketAddr, SocketAddr), UpstreamTx>;
+type Result<T, E = Error> = std::result::Result<T, E>;
+pub type UpstreamTx = mpsc::Sender<Vec<u8>>;
+pub type UpstreamRx = mpsc::Receiver<Vec<u8>>;
+pub type DistributorTx = mpsc::Sender<(SocketAddr, Endpoint, Vec<u8>)>;
+pub type DistributorRx = mpsc::Receiver<(SocketAddr, Endpoint, Vec<u8>)>;
+
+/// UPSTREAM_EXPIRY_INTERVAL is the default interval to check for expired sessions.
+const UPSTREAM_EXPIRY_INTERVAL: Duration = Duration::from_secs(60);
+pub const DEFAULT_TTL: Duration = Duration::from_secs(60);
+
+fn upstream_queue() -> (UpstreamTx, UpstreamRx) {
+    mpsc::channel(1024)
 }
 
-/// ReceivedPacketContext contains state needed to process a received packet.
-struct ReceivedPacketContext<'a> {
-    packet: &'a [u8],
-    filter_manager: SharedFilterManager,
-    endpoint: &'a Endpoint,
-    from: SocketAddr,
-    to: SocketAddr,
+fn distributor_queue() -> (DistributorTx, DistributorRx) {
+    mpsc::channel(1024)
 }
 
-/// Packet represents a packet that needs to go somewhere
-pub struct Packet {
-    dest: SocketAddr,
-    contents: Vec<u8>,
+pub(crate) struct UpstreamDistributor {
+    pub(crate) log: slog::Logger,
+    pub(crate) metrics: Metrics,
+    pub(crate) downstream_socket: Arc<UdpSocket>,
+    pub(crate) filter_manager: SharedFilterManager,
 }
 
-impl Packet {
-    pub fn new(dest: SocketAddr, contents: Vec<u8>) -> Packet {
-        Packet { dest, contents }
-    }
+impl UpstreamDistributor {
+    pub fn spawn(
+        self,
+        mut shutdown_rx: ShutdownRx,
+    ) -> (DistributorTx, JoinHandle<Result<()>>) {
+        let manager: dashmap::DashMap<(SocketAddr, SocketAddr), UpstreamTx> =
```

Comment:
the issue with dashmap is that it isn't async-aware so that any contention would block the waiting threads rather than the current impl that yields (though the current impl locks the whole map). I'm wondering if this is the lesser evil though but not sure what alternatives are available or if we can write something similar ourselves and tackle this in a separate PR? Ideally we want to use a sharded map that yields instead.

Reply:
Yeah, though I believe it should only block on the upstream distributor thread, though your comment made me realise that this whole problem is actually best solved by instead using an LRU cache here, and there are couple of async libraries for that.

Reply:
I don't think a plain lru cache will do here (at least not without redefining what session expiration means), we'd need to limit the size of the map (which could be a nice thing to have configurable in any case) but don't necessarily want to remove older ones when it gets full and won't be able to delete things when they're no longer in use (i.e the cache will remain full even if all sessions are no longer active)

Reply:
> we'd need to limit the size of the map

Well what I was thinking was just setting the capacity of the to `usize` to `usize::MAX` which effectively would give the cache the same capacity as the map and make it so nothing would expired by size, and it would only really be removing based on expiry.

However I realised, that the locking here is actually needless, because we have unique access to the map in the distributor thread, we don't actually need to provide any kind of locking, so now it just uses `HashMap`.